### ISSUE TITLE
More informative exception text for failing pre/post conditions using pdp.cond.Condition objects

### ISSUE THREAD

**Issue by shaypal5** [pdpipe/pdpipe#68]
If I'm using a `pdp.cond.Condition` object, for example `pdp.cond.HasAllColumns(['a', 'b'])`, as a precondition (or a post-condition) in my pipeline stage and it fails, I expect an informative exception message, such as `"Not all required columns ['a', 'b'] found in input dataframe"` (or better yet, `"Required columns ['a'] not found in input dataframe"`, in the case only `"a"` is missing).

This is something that can be achieved by using custom descriptions of specific `Condition` classes, as they provide semantic information on why they failed (or at least, they could).

At the moment, I get a generic `"Precondition failed in stage..."`. :(

**Comment by yarkhinephyo** [comment]
Hi Shy,

For displaying an error message specific to a condition, (Example - `"Not all required columns ['a', 'b'] found in input dataframe"` when only `['a']` is missing), I am thinking add a new attribute to `Condition` to store `self.exdesc` which is populated during object construction by subclass. Then inside `PdPipelineStage`, if `self._prec_arg` returns False during precondition check, ``self._prec_arg.exdesc` is accessed as the string for display.

For displaying a more specific error message, (Example - `"Required columns ['a] found in input dataframe"` when only `['a']` is missing), I am thinking the `def __call__` of condition functions such as `class _NoColumnsFunc`, can return `ConditionResponse` object which contains a boolean and a string instead of just a boolean.

Does it align with what you have in mind?

**Comment by shaypal5** [comment]
I'll say let's start with the more basic type of prints, because thinking about this again it seems like most stages will have to be re-written with a more complex and less efficient code to be able to tell this.

And the simpler implementation can just have a custom attribute that each condition must populate, with the `Condition` class providing a default string.

I would go with a name that shows this attribute is part of the api, so `Condition.error_message`.

Then, the different methods of `PipelineStage` using pre and post conditions can do (for example `fit`):
```python
def fit(self, X, y=None, exraise=None, verbose=False):
    if exraise is None:
        exraise = self._exraise
    if self._compound_prec(X):
        if verbose:
            msg = '- ' + '\n  '.join(textwrap.wrap(self._appmsg))
            print(msg, flush=True)
        res_df = self._fit_transform(X, verbose=verbose)
        if exraise and not self._compound_post(df=res_df):
            try:
                raise FailedPostconditionError(self._compound_post.error_message)
            except AttributeError:
                raise FailedPostconditionError(self._exmsg_post)
        return X
    if exraise:
        try:
            raise FailedPreconditionError(self._compound_prec.error_message)
        except AttributeError:
            raise FailedPreconditionError(self._exmsg)
    return X
```

**Comment by shaypal5** [comment]
We have to do this because a condition can be any `callable` object that can return a `bool` value for `pandas.DataFrame` objects, and might not have any of these attributes.

**Comment by yarkhinephyo** [comment]
Thanks for the detailed explanation!

Just to check again, `Condition` class should look something like this:
```
class Condition(object):
    def __init__(self, func, error_message=None, fittable=None):
        self._func = func
        self._fittable = fittable
        if error_message is not None:
            self.error_message = error_message
...
```

Also would it be more helpful for user if we give them the `"Precondition failed in stage..."` either way? 
Something along this line -

```
if exraise:
    try:
        raise FailedPreconditionError(self._exmsg + self._compound_prec.error_message)
    except AttributeError:
        raise FailedPreconditionError(self._exmsg)

```

**Comment by shaypal5** [comment]
Yes, just add the `error_message` param *after* `fittable`, so argument order is backward compatible.

Regarding 2: Yeah, good idea. Also, a fix to my mistake: It should be `self._prec_arg.error_message`. `_compound_prec` is a method of `PdPipelineStage`, not the `Condition` object that might have been supplied.

**Comment by jbbqqf** [comment]
Hi @shaypal5! Triaging older issues this week — I think this one has shipped and can be closed.

The exact API discussed in the 2021-11-25 thread (`Condition.error_message` constructor parameter, kept after `fittable` for backward compatibility) is in `main` today, plus the wiring to use it:

- [`src/pdpipe/cond.py:38-42`](https://github.com/pdpipe/pdpipe/blob/master/src/pdpipe/cond.py#L38-L42) — `Condition.__init__(self, func, fittable=None, error_message=None)` stores it as `self._error_message`.
- [`src/pdpipe/core.py:665-681`](https://github.com/pdpipe/pdpipe/blob/master/src/pdpipe/core.py#L665-L681) — `_raise_precondition_error` / `_raise_postcondition_error` read `self._prec_arg._error_message` and raise `FailedPreconditionError(error_message)` instead of the generic message. Falls back to `_exmsg` when the attribute is missing, so callable-only conditions still work.
- All built-in conditions populate specific strings — `HasAllColumns` ("Not all required columns [...] present in the input dataframe."), `HasNoColumns`, `HasAtMostMissingValues`, `HasNoMissingValues`, `HasAtMostNQualifyingColumns`, `HasAtLeastNQualifyingColumns`, `HasNoQualifyingColumns` — see [`src/pdpipe/cond.py`](https://github.com/pdpipe/pdpipe/blob/master/src/pdpipe/cond.py).

Quick check:

```python
import pandas as pd, pdpipe as pdp
df = pd.DataFrame({'a': [1, 2]})
stage = pdp.ColDrop('b', prec=pdp.cond.HasAllColumns(['a', 'b']))
stage(df)
# pdpipe.exceptions.FailedPreconditionError: Not all required columns ['a', 'b'] present in the input dataframe.
```

Would you mind closing this out? If something specific is still missing — say a sub-field of the message, or the way it's wired through compound conditions — point me at it and I'll dig further.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against the current `master` branch at `3140064`.)

**Comment by shaypal5** [comment]
Closing as shipped. Current master includes Condition(error_message=...), pre/post-condition error wiring in PdPipelineStage, built-in condition messages, and regression coverage. If there is still a specific missing compound-condition behavior, we can track that in a narrower follow-up.
